### PR TITLE
Fix: Index structure map[string]string,Mongo resulting in inconsistent results obtained by filtering non-string type by index.

### LINKED ISSUE #5263: [Bug] index structure  map[string]string ,mongo  resulting in inconsistent results obtained by filtering non-string type by index.
**Describe the bug**

Now the datastore index structure is map[string]string, the value of non-string type stored in mongo is not string, resulting in inconsistent results obtained by filtering non-string type by index. Can I change the index to map[string]interface? I can submit the relevant fix code.
image
image

**To Reproduce**

**Expected behavior**

**Screenshots**

**KubeVela Version**

**Cluster information**

**Additional context**

### PR DESCRIPTION
**Pull request by LiuFang07**
Signed-off-by: old.prince <di7zhang@gmail.com>

### Description of your changes

Fixes #5263

Value of non-string type stored in mongo is not string.Resulting in inconsistent results obtained by filtering non-string type by index.The reason is that when mongo stores the index is still filtered by string value but the actual value is not a string.

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### CODE REVIEW

File: pkg/apiserver/infrastructure/datastore/kubeapi/kubeapi.go [hunk 1]
```diff
@@ -490,7 +493,24 @@ func (m *kubeapi) Count(ctx context.Context, entity datastore.Entity, filterOpti
...
+		return nil
+	}
+	ret := make(map[string]string, len(index))
+	for k, v := range index {
+		value := pkgUtils.ToString(v)
+		if value == "" {
+			klog.Warningf("unable to cast %#v of type %T to string", v, v)
+			continue
+		}
+		ret[k] = pkgUtils.ToString(v)
```
Comment (StevenLeiZhang) [inline pkg/apiserver/infrastructure/datastore/kubeapi/kubeapi.go:513]: need check the length of Key and Value, make sure there length is less then K8S restriction.
If Key or Value is too long, what is solution?

File: pkg/apiserver/infrastructure/datastore/kubeapi/kubeapi.go [hunk 2]
```diff
@@ -490,7 +493,24 @@ func (m *kubeapi) Count(ctx context.Context, entity datastore.Entity, filterOpti
 }
 
 func verifyValue(v string) string {
+
 	s := strings.ReplaceAll(v, "@", "-")
 	s = strings.ReplaceAll(s, " ", "-")
 	return strings.ToLower(s)
 }
+
+func convertIndex2Labels(index map[string]interface{}) map[string]string {
```
Comment (StevenLeiZhang) [inline pkg/apiserver/infrastructure/datastore/kubeapi/kubeapi.go:502]: Please add test for this function.

File: pkg/utils/strings_test.go [hunk 3]
```diff
@@ -137,3 +137,16 @@ func TestJoinURL(t *testing.T) {
 	}
 
 }
+
+func TestToString(t *testing.T) {
```
Comment (StevenLeiZhang) [inline pkg/utils/strings_test.go:141]: need more test case to import test coverage.

File: pkg/apiserver/infrastructure/datastore/kubeapi/kubeapi.go (hunk 1 above)
Comment (LiuFang07) [reply pkg/apiserver/infrastructure/datastore/kubeapi/kubeapi.go:513]: Index will convert to k8s labels, k8s labels has len limited.The current version of vela does nothing,If necessary can open a feat or fix separately to solve the problem

File: pkg/apiserver/infrastructure/datastore/kubeapi/kubeapi.go [hunk 4]
```diff
@@ -25,6 +25,11 @@ import (
 	"strings"
 	"time"
 
+	pkgUtils "github.com/oam-dev/kubevela/pkg/utils"
```
Comment (barnettZQG) [inline pkg/apiserver/infrastructure/datastore/kubeapi/kubeapi.go:28]: Please keep the code style, and move the inner package to last.

Comment (LiuFang07) [reply pkg/apiserver/infrastructure/datastore/kubeapi/kubeapi.go]: @barnettZQG  ok,modify completed

**Review by barnettZQG — APPROVED** [review]

**Pull request merged**